mini-projects/01-device-driver/laptop-interface/src/serial.rs

```diff
@@ -0,0 +1,159 @@
+    sp_set_bits(port_handle, 8);
+    sp_set_parity(port_handle, SpParity::SP_PARITY_NONE);
+    sp_set_stopbits(port_handle, 1);
+    sp_set_flowcontrol(port_handle, SpFlowcontrol::SP_FLOWCONTROL_NONE);
+
+    // Write "stop controller" to the port to make sure it's in the right state
+    let stop_controller = b"stop controller\n";
+    match blocking_write(port_handle, stop_controller) {
+        Ok(_) => println!(
+            "Wrote {} to the port {}",
+            std::str::from_utf8(stop_controller).unwrap(),
+            port_name
+        ),
+        Err(e) => eprintln!(
+            "Failed to write {} to {}: {}",
+            std::str::from_utf8(stop_controller).unwrap(),
+            port_name,
+            e
+        ),
+    }
+
+    thread::sleep(std::time::Duration::from_millis(100));
+
+    // Write "reset" to the port
+    let reset_controller = b"reset\n";
+    match blocking_write(port_handle, reset_controller) {
+        Ok(_) => println!(
+            "Wrote {} to the port {}",
+            std::str::from_utf8(reset_controller).unwrap(),
+            port_name
+        ),
+        Err(e) => eprintln!(
+            "Failed to write {} to {}: {}",
+            std::str::from_utf8(reset_controller).unwrap(),
+            port_name,
+            e
+        ),
+    }
+
+    thread::sleep(std::time::Duration::from_millis(100));
+
+    // Write "init controller" to the port
+    let init_controller = b"init controller\n";
+    match blocking_write(port_handle, init_controller) {
+        Ok(_) => println!(
+            "Wrote {} to the port {}",
+            std::str::from_utf8(init_controller).unwrap(),
+            port_name
+        ),
+        Err(e) => eprintln!(
+            "Failed to write {} to {}: {}",
+            std::str::from_utf8(init_controller).unwrap(),
+            port_name,
+            e
+        ),
+    }
+
+    Ok(port_handle)
+}
+
+/// Blocking write to the serial port
+pub unsafe fn blocking_write(port_handle: *mut SpPort, data: &[u8]) -> io::Result<usize> {
+    let mut written = 0;
+    while written < data.len() {
+        let bytes_written = sp_blocking_write(
+            port_handle,
+            data.as_ptr() as *const c_void,
+            data.len() as size_t,
+            1000,
+        );
+        if bytes_written < 0 {
+            return Err(io::Error::last_os_error());
+        }
+        written += bytes_written as usize;
+    }
+    Ok(written)
+}
+
+/// Read the state of the controller
+pub unsafe fn read_controller_state(port_handle: *mut SpPort) -> Option<ControllerState> {
+    let mut buffer = [0u8; 3];
+    let bytes_read = sp_blocking_read(
+        port_handle,
+        buffer.as_mut_ptr() as *mut c_void,
+        buffer.len(),
+        1000,
+    );
+
+    if bytes_read == 3 {
+        let adc_value = u16::from_be_bytes([buffer[1], buffer[2]]);
+        Some(ControllerState {
+            buttons: buffer[0],
+            adc_value,
+        })
+    } else {
+        None
```

Comment:
any reason why you opted to not error out if an invalid serial command was received?

Reply:
making an explicit error make signify incompatible device firmware, which would be good to catch rather than silently erroring out on